spec/unit/ssl/state_machine_spec.rb

```diff
@@ -0,0 +1,129 @@
+require 'spec_helper'
+require 'puppet_spec/files'
+
+require 'puppet/ssl'
+
+describe Puppet::SSL::StateMachine do
+  include PuppetSpec::Files
+
+  let(:machine) { described_class.new }
+  let(:cacert_pem) { File.read(fixtures('unit/ssl/ssl_provider/ca.pem')) }
+  let(:cacert) { OpenSSL::X509::Certificate.new(cacert_pem) }
+  let(:cacerts) { [cacert] }
+
+  let(:crl_pem) { File.read(fixtures('unit/ssl/ssl_provider/crl.pem')) }
+  let(:crl) { OpenSSL::X509::CRL.new(crl_pem) }
+  let(:crls) { [crl] }
+
+  context 'when ensuring CA certs and CRLs' do
+    it 'returns an SSLContext with the loaded CA certs and CRLs' do
+      Puppet::X509::CertProvider.any_instance.stubs(:load_cacerts).returns(cacerts)
+      Puppet::X509::CertProvider.any_instance.stubs(:load_crls).returns(crls)
+
+      ssl_context = machine.ensure_ca_certificates
+
+      expect(ssl_context[:trusted_certs]).to eq(cacerts)
+      expect(ssl_context[:crls]).to eq(crls)
+      expect(ssl_context[:verify_peer]).to eq(true)
+    end
+  end
+
+  context 'NeedCACerts' do
+    let(:state) { Puppet::SSL::StateMachine::NeedCACerts.new }
+
+    before :each do
+      Puppet[:localcacert] = tmpfile('needcacerts')
+    end
+
+    it 'transitions to NeedCRLs state' do
+      Puppet::X509::CertProvider.any_instance.stubs(:load_cacerts).returns(cacerts)
+
+      expect(state.next_state).to be_an_instance_of(Puppet::SSL::StateMachine::NeedCRLs)
+    end
+
+    it 'loads existing CA certs' do
+      Puppet::X509::CertProvider.any_instance.stubs(:load_cacerts).returns(cacerts)
+
+      st = state.next_state
+      expect(st.ssl_context[:trusted_certs]).to eq(cacerts)
+    end
+
+    it 'fetches and saves CA certs' do
+      Puppet::X509::CertProvider.any_instance.stubs(:load_cacerts).returns(nil)
+      Puppet::SSL::Fetcher.any_instance.stubs(:fetch_cacerts).returns(cacert_pem)
+
+      st = state.next_state
+      expect(st.ssl_context[:trusted_certs].map(&:to_pem)).to eq(cacerts.map(&:to_pem))
+      expect(File).to be_exist(Puppet[:localcacert])
+    end
+
+    it 'does not verify the peer cert if there are no local CA certs' do
+      Puppet::X509::CertProvider.any_instance.stubs(:load_cacerts).returns(nil)
+
+      Puppet::Rest::Routes.expects(:get_certificate).with do |_, ssl_context|
+        expect(ssl_context[:verify_peer]).to eq(false)
+      end.returns(cacert_pem)
+
+      state.next_state
+    end
+
+    it 'raises if CA certs are invalid' do
+      Puppet::X509::CertProvider.any_instance.stubs(:load_cacerts).returns(nil)
+      Puppet::SSL::Fetcher.any_instance.stubs(:fetch_cacerts).returns('')
+
+      expect {
+        state.next_state
+      }.to raise_error(OpenSSL::X509::CertificateError)
```

Comment:
Should we have a test that shows it doesn't save if it raises?

Reply:
yes, let me add that to this test and the CRL one below too

Reply:
Done